### PR TITLE
[flash_ctrl,dv] more reset tests

### PR DESCRIPTION
**Pull request by jdonjdon**
- reset with rma test
- reset with otp test
Signed-off-by: Jaedon Kim <jdonjdon@google.com>

### CODE REVIEW

**Comment by jdonjdon** [comment]
@tjaychen  Can you expose (https://github.com/lowRISC/opentitan/blob/bee125246928489c7a9dac012c456aa04dd2e483/hw/ip/flash_ctrl/rtl/flash_ctrl_lcmgr.sv#L130-L142)
to flash_ctrl_package?

**Comment by tjaychen** [comment]
> @tjaychen Can you expose (
> 
> https://github.com/lowRISC/opentitan/blob/bee125246928489c7a9dac012c456aa04dd2e483/hw/ip/flash_ctrl/rtl/flash_ctrl_lcmgr.sv#L130-L142
> 
> )
> to flash_ctrl_package?

ah yes let me do that.

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_hw_rma_reset_vseq.sv [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+// Run rma request event with random resets
+class flash_ctrl_hw_rma_reset_vseq extends flash_ctrl_hw_rma_vseq;
+  typedef enum {DVStRmaPageSel     = 0,
+                DVStRmaErase       = 1,
+                DVStRmaEraseWait   = 2,
+                DVStRmaWordSel     = 3,
+                DVStRmaProgram     = 4,
+                DVStRmaProgramWait = 5,
+                DVStRmaRdVerify    = 6} reset_state_index_e;
+  reset_state_index_e reset_state_index;
```
Comment (cindychip) [inline hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_hw_rma_reset_vseq.sv:14]: This does not need to be a global variable right, maybe you can just declare it at line 36.
But just a comment, I have no strong preference.

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_hw_rma_reset_vseq.sv [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+                           "Timed out waiting for rma_state == StRmaProgramWait",
+                           state_wait_timeout_ns)
+            end
+            default: begin
+              `DV_SPINWAIT(wait(cfg.flash_ctrl_dv_vif.rma_state == StRmaRdVerify);,
+                           "Timed out waiting for rma_state == StRmaRdVerify",
+                           state_wait_timeout_ns)
+              cfg.clk_rst_vif.wait_clks(10);
+            end
+          endcase
```
Comment (cindychip) [inline hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_hw_rma_reset_vseq.sv:75]: Can we simplify as below:
```
begin
  reset_state_index_e reset_state_index;
  reset_state_index = $urandom_range(DVStRmaPageSel, DVStRmaRdVerify);
  `uvm_info("Test", $sformatf("Reset index: %0d", reset_state_index), UVM_LOW)
  `DV_SPINWAIT(wait(cfg.flash_ctrl_dv_vif.rma_state == reset_state_index);,
                 $sformatf("Timed out waiting for rma_state: %0s", reset_state_index.name),
                 state_wait_timeout_ns)
  // Maybe give a brief explanation why we have to wait 10 clk cycles here
  if (reset_state_index inside {StRmaRdVerify, StRmaErase}) cfg.clk_rst_vif.wait_clks(10);
   `uvm_info(`gfn, "RESET", UVM_LOW)
  lc_ctrl_if_rst();  // Restore lc_ctrl_if to Reset Values
  cfg.seq_cfg.disable_flash_init = 1;  // Disable Flash Random Initialisation
  apply_reset();
end
```

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_otp_reset_vseq.sv [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    StReqDataKey    = 11'b01110010001,
+    StReadSeeds     = 11'b11011111110,
+    StReadEval      = 11'b01000100111,
+    StWait          = 11'b00100111011,
+    StEntropyReseed = 11'b00011000110,
+    StRmaWipe       = 11'b10010110101,
+    StRmaRsp        = 11'b10110001010,
+    StDisabled      = 11'b11111100011,
+    StInvalid       = 11'b11101011000
+  } lcmgr_state_e;
```
Comment (cindychip) [inline hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_otp_reset_vseq.sv:21]: super nit:
it is called a lc_ctrl :) so maybe `lc_ctrl_state_e`

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_otp_reset_vseq.sv [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        begin
+          csr_wr(.ptr(ral.init), .value(1));
+          `uvm_info("Test","OTP",UVM_LOW)
+          otp_model();
+          `DV_SPINWAIT(wait(cfg.flash_ctrl_dv_vif.rd_buf_en == 1);,
+                       "Timed out waiting for rd_buf_en",
+                       state_wait_timeout_ns)
+        end
+        begin
+          case (reset_index)
```
Comment (cindychip) [inline hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_otp_reset_vseq.sv:45]: same comments for above if that sounds reasonable.

File: hw/ip/flash_ctrl/dv/flash_ctrl_base_sim_cfg.hjson [hunk 5]
```diff
@@ -114,6 +114,18 @@
       run_opts: ["+test_timeout_ns=300000000000"]
       reseed: 3
     }
+    {
+      name: flash_ctrl_hw_rma_reset
+      uvm_test_seq: flash_ctrl_hw_rma_reset_vseq
+      run_opts: ["+test_timeout_ns=300000000000"]
```
Comment (cindychip) [inline hw/ip/flash_ctrl/dv/flash_ctrl_base_sim_cfg.hjson:120]: nit: 
```suggestion
      run_opts: ["+test_timeout_ns=300_000_000_000"]
```

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_hw_rma_reset_vseq.sv (hunk 1 above)
Comment (jdonjdon) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_hw_rma_reset_vseq.sv:14]: I have urandom of reset_state_index at line 25

Comment (jdonjdon) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_hw_rma_reset_vseq.sv:14]: Oh wait, I didn't see you other comment below.

Comment (jdonjdon) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_hw_rma_reset_vseq.sv]: fixed

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_hw_rma_reset_vseq.sv (hunk 2 above)
Comment (jdonjdon) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_hw_rma_reset_vseq.sv]: That's great. Rewrote as suggested.

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_otp_reset_vseq.sv (hunk 3 above)
Comment (jdonjdon) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_otp_reset_vseq.sv]: fixed

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_otp_reset_vseq.sv (hunk 4 above)
Comment (jdonjdon) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_otp_reset_vseq.sv]: fixed

File: hw/ip/flash_ctrl/dv/flash_ctrl_base_sim_cfg.hjson (hunk 5 above)
Comment (jdonjdon) [reply hw/ip/flash_ctrl/dv/flash_ctrl_base_sim_cfg.hjson]: fixed

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_hw_rma_reset_vseq.sv [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+          // Give more cycles for long stages
+          // to trigger reset in the middle of the state.
+          if (reset_state_index inside {StRmaRdVerify, StRmaErase}) cfg.clk_rst_vif.wait_clks(10);
+          `uvm_info(`gfn, "RESET", UVM_LOW)
+          lc_ctrl_if_rst();  // Restore lc_ctrl_if to Reset Values
+          cfg.seq_cfg.disable_flash_init = 1;  // Disable Flash Random Initialisation
+          apply_reset();
+        end
+      join_any
+      disable fork;
```
Comment (cindychip) [inline hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_hw_rma_reset_vseq.sv:48]: Another thing I was thinking about this multi-thread.
Will there be cases that `disable fork` happened during `apply_reset` that we ended up stuck at reset inactive?
Maybe a better way would be applying reset after `disable fork` if the reset condition is satisified?

Comment (jdonjdon) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_hw_rma_reset_vseq.sv:48]: > Another thing I was thinking about this multi-thread. Will there be cases that `disable fork` happened during `apply_reset` that we ended up stuck at reset inactive? 
=> No. Because all these states are sub state of  RMA.
So the 2nd begin/end always finish before send_rma_req() ends.

This seq and another (flash_ctrl_otp_reset_vseq) is apply reset during non-usual operation.
The reason I created directed test rather than using our random_reset  sequence is,
some state is just one cycle and others are a few msec. So with 'random_reset' sequence, it is  really hard
to hit each corner case.

Comment (cindychip) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_hw_rma_reset_vseq.sv:48]: Good to know, do you mind - if not too late, add the comments there... I am sure a few months later, I will forget and ask you the same question again haha.

Comment (jdonjdon) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_hw_rma_reset_vseq.sv:48]: > Good to know, do you mind - if not too late, add the comments there... I am sure a few months later, I will forget and ask you the same question again haha.

No problem. Added comment on both files. :)

**Review by cindychip — APPROVED** [review]

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_hw_rma_reset_vseq.sv [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+      // Since the 2nd begin/end wait for substate of RMA,
+      // the 2nd begin/end always finish first.
+      // Soi diable fork only termintate send_rma_req.
```
Comment (cindychip) [inline hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_hw_rma_reset_vseq.sv:51]:
```suggestion
      // So disable fork only terminate send_rma_req.
```

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_otp_reset_vseq.sv [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+                       $sformatf("Timed out waiting for %s", reset_index.name),
+                       state_wait_timeout_ns)
+
+          // Add extra cycles for longer states
+          // to trigger reset in the middle of the state.
+          if (reset_index == DVWaitAddrKey || reset_index == DVWaitDataKey) begin
+              cfg.clk_rst_vif.wait_clks(2);
+          end else if (reset_index == DVWaitReadSeeds || reset_index == DVWait) begin
+              cfg.clk_rst_vif.wait_clks(10);
+          end
```
Comment (cindychip) [inline hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_otp_reset_vseq.sv:53]: nit:
```suggestion
          if (reset_index == DVWaitAddrKey || reset_index == DVWaitDataKey) begin
            cfg.clk_rst_vif.wait_clks(2);
          end else if (reset_index == DVWaitReadSeeds || reset_index == DVWait) begin
            cfg.clk_rst_vif.wait_clks(10);
          end
```

File: hw/ip/flash_ctrl/dv/flash_ctrl_base_sim_cfg.hjson [hunk 9]
```diff
@@ -114,6 +114,18 @@
...
+    {
+      name: flash_ctrl_hw_rma_reset
+      uvm_test_seq: flash_ctrl_hw_rma_reset_vseq
+      run_opts: ["+test_timeout_ns=300_000_000_000"]
+      reseed: 20
+    }
+    {
+      name: flash_ctrl_otp_reset
+      uvm_test_seq: flash_ctrl_otp_reset_vseq
+      run_opts: ["+test_timeout_ns=300000000000"]
```
Comment (cindychip) [inline hw/ip/flash_ctrl/dv/flash_ctrl_base_sim_cfg.hjson:126]: nit:
```suggestion
      run_opts: ["+test_timeout_ns=300_000_000_000"]
```

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_hw_rma_reset_vseq.sv (hunk 7 above)
Comment (jdonjdon) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_hw_rma_reset_vseq.sv]: whoops. fixed

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_otp_reset_vseq.sv (hunk 8 above)
Comment (jdonjdon) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_otp_reset_vseq.sv:53]: fixed

File: hw/ip/flash_ctrl/dv/flash_ctrl_base_sim_cfg.hjson (hunk 9 above)
Comment (jdonjdon) [reply hw/ip/flash_ctrl/dv/flash_ctrl_base_sim_cfg.hjson:126]: fixed.
